### PR TITLE
Fix `DM_DEFAULT_ENCODING` SpotBugs violations

### PR DESCRIPTION
**Pull request by basil**
Fix `DM_DEFAULT_ENCODING` SpotBugs violations by explicitly passing in the default character set where possible (or `StandardCharsets.UTF_8` for Unix-specific functionality).

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/AbstractBuild.java
Comment (basil) [inline core/src/main/java/hudson/model/AbstractBuild.java:672]: The change to `SCM.java` elsewhere in this PR (from traditional Java I/O to NIO.2) means that cancelling a job while SCM checkout is taking place can now throw `ClosedByInterruptException` as well as `InterruptedException` and `InterruptedIOException`. Higher-level code marks the build as `ABORTED` for `InterruptedException` and `FAILURE` for other I/O exceptions, so we convert the `ClosedByInterruptException` to an `InterruptedException` here (just as the existing code does for `InterruptedIOException`) so that the build will be shown as `ABORTED` (as with the pre-NIO.2 logic).

**Review by oleg-nenashev — APPROVED** [review]
All changes look reasonable. PEMs files should always be UTF-8, same as Unix process files IIRC

**Review by jtnord — COMMENTED** [review]
At least one spotbugs warning was masking a real issue.

I have looked at the other ones and can not say for certain that all of them are correct but they appear to be.   Given this change would then mask the reports of potential issues it may be better to have a few more eyes on this than normal.  
To be clear - this change does not introduce any bug - but as the spotbugs warnings are fixed, there is no longer a report to look into something.

**Review by jtnord — APPROVED** [review]
I can not say for certain that the changes are all that all correct but they appear to be. Given this change would then mask the reports of potential issues it may be better to have a few more eyes on this than normal.

To be clear - this change does not introduce any bug - but as the spotbugs warnings are fixed, there is no longer a report to look into something.

**Comment by basil** [comment]
> it may be better to have a few more eyes on this than normal.

Any other takers?

File: core/src/main/java/hudson/model/Run.java
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java:1516]: I think this is OK. https://github.com/jenkinsci/jep/blob/master/jep/206/README.adoc means that for Pipeline this should always be UTF-8.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateCenter.java:2028]: Uh, let us use ASCII here perhaps.

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateCenter.java:2032]: ditto

File: core/src/main/java/hudson/scm/SCM.java
Comment (jglick) [inline core/src/main/java/hudson/scm/SCM.java:744]: Hmm, I guess we do need to use platform charset for compatibility here. (In most cases the file is _not_ XML despite appearances.)

File: core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java:104]: UTF-8

File: core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java
Comment (jglick) [inline core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java:88]: ASCII or UTF-8

File: core/src/main/java/hudson/util/CompressedFile.java
Comment (jglick) [inline core/src/main/java/hudson/util/CompressedFile.java:119]: Maybe? No idea what this class is even used for.

File: core/src/main/java/hudson/util/LogTaskListener.java
Comment (jglick) [inline core/src/main/java/hudson/util/LogTaskListener.java:93]: UTF-8

File: core/src/main/java/jenkins/security/DefaultConfidentialStore.java
Comment (jglick) [inline core/src/main/java/jenkins/security/DefaultConfidentialStore.java:155]: ASCII or UTF-8

**Review by jglick — CHANGES_REQUESTED** [review]
A number of places here where default charset is specified but I think UTF-8 would be more correct. I understand that you are actually making the more conservative change here—retaining existing behavior—but by making the warning go away and adding an explicit call to `defaultCharset` it makes it look like this was intentional, when generally it was not, thus making it less likely that the code would be corrected in the future compared to just leaving it alone (or adding a suppression annotation).

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (basil) [reply core/src/main/java/hudson/cli/CLICommand.java:409]: Done

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (basil) [reply core/src/main/java/hudson/cli/CLICommand.java:423]: Done

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (basil) [reply core/src/main/java/hudson/cli/CLICommand.java:440]: Done

File: core/src/main/java/hudson/console/ConsoleNote.java
Comment (basil) [reply core/src/main/java/hudson/console/ConsoleNote.java]: Done

File: core/src/main/java/hudson/console/ConsoleNote.java
Comment (basil) [reply core/src/main/java/hudson/console/ConsoleNote.java]: Done

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (basil) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Done

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (basil) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Done

File: core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java
Comment (basil) [reply core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java:104]: Done

File: core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java
Comment (basil) [reply core/src/main/java/hudson/security/csrf/DefaultCrumbIssuer.java]: The buffer can contain an authentication principal, so let's use UTF-8 for that. The salt is a hexadecimal string, so ASCII should suffice.

File: core/src/main/java/hudson/util/CompressedFile.java
Comment (basil) [reply core/src/main/java/hudson/util/CompressedFile.java:119]: It's unused. I deprecated it and will remove in a few months if no callers show up.

File: core/src/main/java/hudson/util/ConsistentHash.java
Comment (basil) [reply core/src/main/java/hudson/util/ConsistentHash.java:308]: Yes, because the hash could include the node name.

File: core/src/main/java/hudson/util/LogTaskListener.java
Comment (basil) [reply core/src/main/java/hudson/util/LogTaskListener.java:93]: Done

File: core/src/main/java/jenkins/security/DefaultConfidentialStore.java
Comment (basil) [reply core/src/main/java/jenkins/security/DefaultConfidentialStore.java:155]: `"::::MAGIC::::"` is valid ASCII, so let's use that.

**Review by jglick — APPROVED** [review]
OK I think.

**Comment by basil** [comment]
> OK I think.

Thanks. I'll run this through BOM testing before merging for good measure.

**Comment by jglick** [comment]
> run this through BOM testing

Would not hurt, though I doubt PCT would catch interesting regressions here since all the CI machines should be using UTF-8 system encoding anyway.

**Comment by basil** [comment]
> Would not hurt, though I doubt PCT would catch interesting regressions here since all the CI machines should be using UTF-8 system encoding anyway.

But re-running the tests after I merge the Windows CI PR should cover that.

**Comment by jglick** [comment]
Yes that should help. I suspect any actual regressions would be in unusual user configurations not covered by tests.

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by jglick** [comment]
> places here where default charset is specified but I think UTF-8 would be more correct

OTOH as of https://openjdk.org/jeps/400 it should not matter typically, so choice of `Charset.defaultCharset()` vs. `StandardCharsets.UTF-8` vs. (where permitted without deprecation warning!) omitting a `Charset` or `String charset` argument would be a matter of style and legibility.

**Pull request merged**
